modules/@apostrophecms/migration/lib/addMissingSchemaFields.js

```diff
@@ -0,0 +1,141 @@
+    // operation
+    addMissingSchemaFieldsFor(doc, schema, dotPath, changes) {
+      for (const field of schema) {
+        const newDotPath = dotPath ? `${dotPath}.${field.name}` : field.name;
+        // Supply the default
+        if (doc[field.name] === undefined) {
+          // Only undefined should fall back here
+          const def = klona((field.def === undefined) ? self.apos.schema.fieldTypes[field.type]?.def : field.def);
+          if (def !== undefined) {
+            if (!Object.hasOwn(changes, dotPath)) {
+              changes[newDotPath] = def;
+            }
+            // Also change it in memory so that if this is a subproperty of a
+            // new object, the change for that new object will have this
+            // subproperty too, plus we don't get crashes above when testing the
+            // subproperties' current values
+            doc[field.name] = def;
+          }
+        }
+        // Address defaults of subproperties
+        if (field.type === 'area') {
+          const basePath = `${newDotPath}.items`;
+          for (let i = 0; (i < (doc[field.name]?.items || []).length); i++) {
+            const widgetPath = `${basePath}.${i}`;
+            const widget = doc[field.name].items[i];
+            const widgetSchema = self.apos.area.getWidgetManager(widget.type)?.schema;
+            if (!widgetSchema) {
+              continue;
+            }
+            self.addMissingSchemaFieldsFor(widget, widgetSchema, widgetPath, changes);
+          }
+        } else if (field.type === 'object') {
+          self.addMissingSchemaFieldsFor(doc[field.name], field.schema, newDotPath, changes);
+        } else if (field.type === 'array') {
+          for (let i = 0; (i < (doc[field.name] || []).length); i++) {
+            const itemPath = `${newDotPath}.${i}`;
+            const item = doc[field.name][i];
+            self.addMissingSchemaFieldsFor(item, field.schema, itemPath, changes);
+          }
+        } else if (field.type === 'relationship') {
+          for (const [ key, item ] of Object.entries(doc[field.fieldsStorage] || {})) {
+            const itemPath = `${newDotPath}.${field.fieldsStorage}.${key}`;
+            self.addMissingSchemaFieldsFor(item, field.schema, itemPath, changes);
+          }
+        }
+      }
+      return changes;
+    },
+    getPropLists() {
+      const schema = {};
+      for (const [ name, module ] of Object.entries(self.apos.doc.managers)) {
+        if (!module.__meta.name) {
+          // Just a placeholder for a type present in the
+          // database but not found in the project code
+          continue;
+        }
+        schema[name] = [];
+        self.expandPropList(module.schema, schema[name], '');
+      }
+      return schema;
+    },
+    expandPropList(schema, propList, dotPath) {
+      for (const field of schema) {
+        const newDotPath = dotPath ? `${dotPath}.${field.name}` : field.name;
```

Comment:
We're not coming up with a real dot path to patch anything at this stage, just a conceptual dot path that namespaces these properties enough to determine if there are differences from one generation of code to the next by comparing two lists of paths.

Real dot paths get figured out in the guts of addMissingSchemaFieldsFor, which does the hard work if a difference does exist.